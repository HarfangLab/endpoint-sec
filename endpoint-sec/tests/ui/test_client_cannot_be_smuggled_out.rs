fn main() {
    use std::cell::RefCell;

    use endpoint_sec::Client;

    Client::new(|client, _| {
        let other_client = Client::new(|_, _| {}).unwrap();
        // Tear the `client` out of the `&mut` reference.
        let smuggled_client = std::mem::replace(client, other_client);
        thread_local! {
            static CLIENT_SMUGGLING: RefCell<Option<Client<'static>>> = RefCell::new(None);
        }
        CLIENT_SMUGGLING.with(|r| *r.borrow_mut() = Some(smuggled_client));
    });
}
