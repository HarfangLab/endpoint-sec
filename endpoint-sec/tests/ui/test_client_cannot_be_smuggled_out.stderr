error[E0521]: borrowed data escapes outside of closure
  --> tests/ui/test_client_cannot_be_smuggled_out.rs:13:9
   |
 6 |     Client::new(|client, _| {
   |                  ------
   |                  |
   |                  `client` is a reference that is only valid in the closure body
   |                  has type `&mut Client<'1>`
...
13 |         CLIENT_SMUGGLING.with(|r| *r.borrow_mut() = Some(smuggled_client));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         `client` escapes the closure body here
   |         argument requires that `'1` must outlive `'static`
   |
   = note: requirement occurs because of the type `RefCell<Option<Client<'_>>>`, which makes the generic argument `Option<Client<'_>>` invariant
   = note: the struct `RefCell<T>` is invariant over the parameter `T`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
